File: 199. Binary Tree Right Side View/solution.py
# Definition for a binary tree node.
from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BFSSolution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        queue = deque([root])
        res = []
        while queue:
            n = len(queue)
            for i in range(n):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                if i == n-1:
                    res.append(node.val)
        return res


class DFSSolution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        ans = []

        def dfs(node=root, level=1):
            if not node:
                return
            if len(ans) < level:
                ans.append(node.val)
            dfs(node.right, level+1)
            dfs(node.left, level+1)
        dfs()
        return ans

File: 199. Binary Tree Right Side View/solution.go
package main

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func dfs(node *TreeNode, level int, res *[]int) {
	if node == nil {
		return
	}
	if len(*res) < level {
		*res = append(*res, node.Val)
	}
	dfs(node.Left, level+1, res)
	dfs(node.Right, level+1, res)
}

func rightSideView(root *TreeNode) []int {
	res := []int{}
	dfs(root, 1, &res)
	return res
}